### DISCUSSION TITLE
Hey model, let's improve you

### DISCUSSION THREAD

**Discussion by da-kami** [get10101/itchysats#1438] [Ideas]
While working on getting "taker goes short" together we noticed that the model has certain weird quirks:

1. Passing around the base values all the time: For many calculations we pass around the base values (i.e. `initial_price`, `quantity`, `leverage`), but we actually only need these values to calculate the `margin`. 
    1. Initially we had a version where we calculated the `margin` and then just store that instead of the base values. 
    2. We should evaluate which values we need at later points. If we only need values to calculate another value we might as well calculated it initially and store the calculated value.
2. `quantity` should be renamed to `contracts`
3. `Order` should be renamed to `Offer`, because it represents the maker's offer to the taker
    1. The offer's `position` should be the one of the taker, because we are offering to the taker to go long/short. Currently the position in the `Order` is the one of the maker (whereas the leverage is the taker's... super confusing). 

With the changes of https://github.com/itchysats/itchysats/pull/1435 there are some new model constraints:

1. The leverage is initially derived from the `Role` (which is derived from the `Origin` of the offer - i.e. if the offer is ours we are the maker, if not we are the taker). The leverage of the `taker` is currently taken out of the `Order`, the maker's always just assumed to be `1`.
   1. Internally we store the `leverage` as `long` and `short` leverage, because it makes it easier to handle calculations. 

## Long/Short and Taker/Maker

I think internally we should always try to depict long vs. short, and *not* taker vs. maker. The only points where taker vs. maker is relevant are:
 
1. In the projection *IF* we want to have a model of "ours" and "theirs" there - this is currently mixed/unclear in many cases (e.g. the "leverage" in the projection cfd is always the taker's leverage)
 2. When we determine the leverage.
 3. When handling certain fees (e.g. opening fee that always goes from taker to maker)
 4. When initially creating / receiving an `Offer` (depending on if we are the maker or taker one of them has to flip the position of the offer to depict what position the Cfd has their point of view)

I wonder if we should change this in `maia` as well, where we keep talking about maker and taker.

**Comment by klochowicz** [comment]
re 3: it's not as easy, seems like maker creates the `Offer`, but in the process of taking it the taker converts it into an `Order` (by adding additional values, such as chosen amount of contracts and leverage). That would be also consistent with our event storming workshop we had some time ago.

**Reply by da-kami** [reply]
I agree, but what is currently called `Order` *is* the `Offer` - the fact that the taker uses that to communicate to the maker what he wants to do is a problem on top of this. I think we should rename the `Order` to `Offer`, and can introduce an `Order` (from taker to maker) in an additional, independent step.

**Reply by klochowicz** [reply]
not quite, because it contains the leverage! (it's *almost* one, I'd grant you that we could rename it, but we should also use a range for leverage, ie. `min_leverage` and `max_leverage`)

**Reply by da-kami** [reply]
I think that notion is wrong. It does not contain the leverage, but the *choices* of leverage that the maker **offers**! The fact that we only have one at the moment and just keep using it for the taker is a simplification that will have to go away once we add different leverages (as you point out yourself - it has to be a range). Then the taker will have to send the decided `quantity` (i.e. contracts) *and* the decided `leverage` to the maker.
So, I think renaming it to `Offer` would be correct even now - it's still confusing from the taker's perspective when taking it though :)
Sidenote: Initially it was called `Offer` for some time until I wrongly renamed it to `Order`. That's one of the rash decisions I regret the most ;)

**Comment by thomaseizinger** [comment]
> Passing around the base values all the time: For many calculations we pass around the base values (i.e. `initial_price`, `quantity`, `leverage`), but we actually only need these values to calculate the `margin`.
> 
>     1. Initially we had a version where we calculated the `margin` and then just store that instead of the base values.

Another thought on this (taking into account combined positions). Here is my current understanding:

When go long, you buy contracts. You buy them at a certain price, which, together with your leverage defines how much margin (collateral) you have to put up. Once you've bought them, the price is irrelevant. You now own these contracts and you have a certain margin locked up.

Your profit/loss depends on the number of contract that you own (which have a value of $1/contract) and the current price.
Your liquidation price depends on the margin you have locked up and how many contracts you own.

If you were to buy more contracts, your position changes:

- You add contracts to it
- You lock up more margin

How much margin you need to lock up depends on the price and leverage **of that second purchase**.

Once the second purchase is completed, you now own more contracts and you have more margin locked up. The new liquidated price depends on both of them (again).

---

My conclusion is: Once you have a CFD, your leverage should actually be _computed_. The source of truth of your position are the number of contracts and your margin, not the other way around. This makes sense once you consider combined positions: Buying more contracts with a different leverage changes the _effective_ leverage of your position AND the liquidation price.

**Reply by da-kami** [reply]
Here is my summary to make sure our understanding is aligned:

* For creating a position for an asset pair we need the leverage of both parties.
* We use leverages to calculate the margins. The margin is the only thing that matters once open. From this point forward the leverage is computed - but will initially be computed as the leverage for opening the position.
* When modifying the position (by buying long or selling short) we need the leverage again, which can be a different leverage. The calculated leverage can then be different than the initial one.
